pkg/monitor/loki_upload.go

```diff
@@ -0,0 +1,285 @@
+		return err
+	}
+
+	logrus.Info("bearer token obtained")
+	client := &http.Client{}
+
+	// NOTE: there is a promtail go client available at https://github.com/grafana/loki/blob/main/clients/pkg/promtail/client/client.go,
+	// however it appears near impossible to vendor. For now, we roll our own with http requests, the worst
+	// part of which is we lose out on resilient retries in the event the ingesters are overloaded.
+
+	values := make([][]interface{}, 0)
+	// This format matches the labels used for the cluster logs in loki, and we want this
+	// to be labelled the same.
+	invoker := fmt.Sprintf("openshift-internal-ci/%s/%s", jobName, buildID)
+	if invoker == "" {
+		return fmt.Errorf("OPENSHIFT_INSTALL_INVOKER is needed for the invoker loki label")
+	}
+	labels := map[string]string{
+		"invoker": invoker,
+		"type":    "origin-interval",
+		// TODO: in future, may group intrvals in origin rather than in the js itself, if we do,
+		// ensure this group gets carried over to this series.
+	}
+
+	for _, i := range intervals {
+		logLine := map[string]string{
+			"_entry": i.Message,
+		}
+
+		if strings.HasPrefix(i.Locator, "e2e-test/\"") {
+			startIndex := strings.Index(i.Locator, "\"") + 1
+			endIndex := strings.LastIndex(i.Locator, "\"")
+			logLine["e2e-test"] = i.Locator[startIndex:endIndex]
+			continue
+		}
+
+		locatorParts := strings.Split(i.Locator, " ")
+
+		for _, lp := range locatorParts {
+			parts := strings.Split(lp, "/")
+			tag, val := parts[0], parts[1]
+			if strings.TrimSpace(tag) == "" {
+				// Have seen this fail on: WARN[0002] unable to process locator tag: May 12 11:10:00.000 I /machine-config reason/OperatorVersionChanged clusteroperator/machine-config-operator version changed from [{operator 4.14.0-0.nightly-2023-05-03-163151}] to [{operator 4.14.0-0.nightly-2023-05-12-121801}]
+				// Would be interesting to track down where this is coming from. Should be identification.go OperatorLocator, but this is impossible.
+				logrus.Warnf("unable to process locator tag: %+v", i)
+				continue
+			}
+
+			// WAARNING: opting for a potentially risky change here, I want namespace filtering to be available as a
+			// label in loki soon,	and thus I am translating some labels we used historically in origin intervals to
+			// match those we pull from pod logs in the cluster itself. This will require our intervals charts in sippy
+			// be able to parse this new name instead of the other. Ideally, we would go rename these fully in origin wherever used.
+			if tag == "ns" {
+				tag = "namespace"
+			}
+			if tag == "node" {
+				tag = "host"
+			}
+
+			logLine[tag] = val
+		}
+
+		// Convert the log data to a packed json log line, where the message is in '_entry'.
+		// This allows us to search additional fields we didn't index as labels using the unpack/json
+		// directives in logQL.
+		logLineJson, err := json.Marshal(logLine)
+		if err != nil {
+			logrus.WithError(err).Errorf("unable to unmashall log line: %s", logLine)
```

Comment:
Do you want to continue here if logLineJson isn't valid or intentionally append the presumably nil logline?

Reply:
Good catch, also should read marshall not unmarshall, thanks!